src/classes/actions-cache/upload.ts

```diff
@@ -0,0 +1,180 @@
+import * as core from '@actions/core';
+import fs from 'fs';
+import {HttpClient} from '@actions/http-client';
+import {TypedResponse} from '@actions/http-client/lib/interfaces';
+import {ReserveCacheError, ValidationError} from '@actions/cache';
+import {isSuccessStatusCode} from './http-responses';
+import {retryHttpClientResponse, retryTypedResponse} from './retry';
+import {getOctokit} from '@actions/github';
+import {retry as octokitRetry} from '@octokit/plugin-retry';
+import {createHttpClient, getCacheApiUrl} from './http-client';
+
+const uploadFile = async (
+  httpClient: HttpClient,
+  cacheId: number,
+  filePath: string,
+  fileSize: number
+): Promise<void> => {
+  if (fileSize <= 0) return;
+  const start = 0;
+  const end = fileSize - 1;
+  const contentRange = `bytes ${start}-${end}/*`;
+  core.debug(
+    `Uploading chunk of size ${
+      end - start + 1
+    } bytes at offset ${start} with content range: ${contentRange}`
+  );
+
+  const additionalHeaders = {
+    'Content-Type': 'application/octet-stream',
+    'Content-Range': contentRange
+  };
+
+  const resourceUrl = getCacheApiUrl(`caches/${cacheId.toString()}`);
+  const fd = fs.openSync(filePath, 'r');
+  try {
+    const uploadChunkResponse = await retryHttpClientResponse(
```

Comment:
Let's move it to the separate function, it's a bit hard to read with all the wraps around. Seems like `uploadChunk` would be an appropriate name for such function.

Reply:
Well, then `uploadFile` will have nothing but call of `uploadChunk` and become useless. 

I improved readability a bit by placing opening a stream in its own function instead.

Reply:
Alright, this way is fine too.